app/resources/admin.html.template

```diff
@@ -68,38 +69,74 @@
-    for (var dropdown_index = 1; $('language_selection_' + dropdown_index) !=
+    for (var dropdown_index = 1; $('repo_titles_language_selection_' + dropdown_index) !=
          null; dropdown_index++) {
-      var language_code = $('language_selection_' + dropdown_index).value;
-      var language_title = $('language_title_' + dropdown_index).value;
+      var language_code = $('repo_titles_language_selection_' + 
+                             dropdown_index).value;
       if (language_code === 'unused') {
         continue;
       }
-      if (arr_contains(used_languages, language_code)) {
-        alert("You entered the same language (" + language_code + ") twice. " +
-              "That's not allowed.");
-        return false;
-      }
       used_languages.push(language_code);
-      languages_to_titles[language_code] = language_title;
     }
 
+    var filtered_index = [];
+    used_languages.filter(function(element, index, self){
+      if (self.indexOf(element) === index){
+        filtered_index.push(index);
+      } else {
+        used_languages.splice(index, 1);
+      }
+    });
+
     // Copy the dropdowns and the titles into the text boxes that will
     // actually be POSTed (language_menu_options and repo_titles).
     $('language_menu_options').value = JSON.stringify(used_languages);
-    $('repo_titles').value = JSON.stringify(languages_to_titles);
+    var languages_to_values = {};
+
+    for (var id_num = 0; id_num < CUSTOM_MESSAGE_TABLE_IDS.length; id_num++){
+      var table_name = CUSTOM_MESSAGE_TABLE_IDS[id_num];
+      for (var i = 0; i < filtered_index.length; i++){
+        var index = filtered_index[i];
+        var language_value = $(table_name + '_language_value_' + (index+1)).value;
+        var used_language = used_languages[index];
+        languages_to_values[used_language] = language_value;
+      }
+      $(table_name).value = JSON.stringify(languages_to_values);
+    }
 
     return true;
   }
 
+  function check_language_selected(row_number) {
+    var language_codes = $('repo_titles_language_selection_' + row_number);
+    var selected_index = language_codes.selectedIndex;
+    var language_code = language_codes[selected_index].value;
+    var used_languages =
+        {{view_config_json.language_menu_options|safe|default:'[]'}};
+    if (used_languages.indexOf(language_code) >= 0) {
+      alert('You entered the same language!!');
+    }
+  }
+
+  function update_language_selected(row_number) {
+    var language_codes = $('repo_titles_language_selection_' + row_number);
+    var selected_index = language_codes.selectedIndex;
+    var language = language_codes[selected_index].text;
+    var language_selected_elements = document.getElementsByClassName(
+                                     'language_selected_' + row_number);
+    for (var i = 0; i < language_selected_elements.length; i++) {
+      language_selected_elements[i].innerHTML = language;
```

Comment:
Can you apply HTML escaping e.g., "<" to "&lt;" to language? It's unlikely that language name contains such characters, but it would be better to prepare for it.